src/LoginCommand.php

```diff
@@ -435,14 +435,12 @@ private function makeMagicUrl(WP_User $user, $expires, $redirect_url)
     private function persistMagicUrl(MagicUrl $magic, $endpoint, $expires)
     {
         if (! set_transient(
-            self::OPTION . '/' . $magic->getKey(),
+            self::OPTION . '/' . wp_hash($magic->getKey()),
```

Comment:
wp_hash is "okay." But it uses md5-based HMAC by default.

https://developer.wordpress.org/reference/functions/wp_hash/#source

And since it's a pluggable function, other implementations could theoretically use something different, which could potentially decrease security if they just make it a simple hash without a secret key.

It would be ideal to use:

https://www.php.net/manual/en/function.sodium-crypto-generichash.php

With wp_salt('auth') as the key.

Reply:
Sure, we can use that since WP polyfills the libsodium library 👍

Reply:
So it seems like `wp_salt('auth')` won't work for the key here because it can generate something longer than the maximum length it supports. I'm getting this error when trying to use it.

```
Uncaught SodiumException: unsupported key length
```

The core [salt generator](https://api.wordpress.org/secret-key/1.1/salt/) produces values that are 64 bytes long, so using these `wp_salt()` will return a 128 byte string. The BLAKE2b algorithm apparently only supports up to a 64 byte key. It's also worth noting that the values for the site's keys and salts can really be anything though, and it's also pluggable like you mentioned.

How would you suggest working around this?

Reply:
Update: perhaps I'm going about this wrong, but simply limiting the key length to 64 bytes does avoid the exception (this is probably not the right way to go about it but as a test), but the resulting string (e.g. `wp_cli_login/�&6m�)v�& ���g��Sϴa Y���Hf�`) doesn't seem to be usable as a transient/option name as it fails to be inserted.

Reply:
Yea, libsodium returns raw bytes. 

For transient names, you should use: 

https://www.php.net/manual/en/function.sodium-bin2base64.php

Probably the URLSAFE enconding is what we want

Reply:
> The core salt generator produces values that are 64 bytes long, so using these wp_salt() will return a 128 byte string. The BLAKE2b algorithm apparently only supports up to a 64 byte key.

True, there a several options: 

Use a custom constant written to wp-config during CLI Install. (I prefer this)

Or you hash the wp salt thus reducing length. 

I don't think that simply substr is good here, since it's a constant time operation.

It might be the case the Core also polyfills:

https://github.com/paragonie/constant_time_encoding

In which case you could use Binary::substr

Reply:
> Yea, libsodium returns raw bytes.
> 
> For transient names, you should use:
> 
> https://www.php.net/manual/en/function.sodium-bin2base64.php

Sounds good, I'll give that a try 👍 

> Use a custom constant written to wp-config during CLI Install. (I prefer this)

We shouldn't write to wp-config because some frameworks manage this via a separate file, so wp-config.php is not meant to be modified directly (e.g. [Roots Bedrock](https://github.com/roots/bedrock/blob/96dbd4e50b7afc0b77b18f10c2dddcf26b29120c/web/wp-config.php)).

The on-install part is also a bit tricky because CLI and server parts can be updated separately. The command will, going forward require the latest version of the server plugin, but that might also be installed via Composer or manually where we wouldn't be able to hook in at that time reliably. If we can find a way to use `wp_salt`, I think that would be best 👍 

> Or you hash the wp salt, thus reducing length.

I think this makes the most sense, but I assume md5/sha1 would still be out of the question here. sha256 would do it, but it looks like it isn't available until PHP 7.1. Of course it would be good to raise the minimum required version of PHP, but the command currently supports the same minimum as WP core (5.6).

Is there another function from libsodium we might use for this?

> I don't think that simply `substr` is good here,.
>
> It might be the case that Core also polyfills:
> 
> https://github.com/paragonie/constant_time_encoding
> 
> In which case, you could use Binary::substr

Is the concern that we'd be using `substr` on a binary string? If so, that shouldn't be a problem because `wp_salt` doesn't return binary. 

I wasn't really thinking we'd use `substr` here, but using standard 64 byte salts it essentially ends up returning `AUTH_KEY` because `wp_salt('auth')` returns `AUTH_KEY . AUTH_SALT`

Reply:
Yes there is another function.

You can call wp_salt(). 

Check if length > 64 (there is constant for this called something like GENERIC_HASH_KEYBYTES_MAX), if greater, 
Use the same sodium_generic_hash function to reduce the length.

Reply:
Should also perform a sanity check to verify that wp_salt() is not:

"put your unique phrase here"

In which case, I'd error.

Reply:
> Check if length > 64 (there is constant for this called something like GENERIC_HASH_KEYBYTES_MAX), if greater,
> Use the same sodium_generic_hash function to reduce the length.

Got it. In this case, what would we use for the key in this case? Use the default empty string?

> Should also perform a sanity check to verify that wp_salt() is not:
> 
> "put your unique phrase here"
> 
> In which case, I'd error.

This sounds like a good idea too. WP also does something like this in the newer recovery mode feature, although in this case it generates a value that is stored in the DB
https://developer.wordpress.org/reference/classes/wp_recovery_mode_cookie_service/recovery_mode_hash/ It also checks for a translated version of the same default value.

Reply:
Correct, to reduce the length you use use sodium_generichash without a key

Reply:
> Should also perform a sanity check to verify that wp_salt() is not:
> 
> "put your unique phrase here"
> 
> In which case, I'd error.

@calvinalkan what would the risk be if we didn't? That is, what if someone did have the default salts and keys? The server plugin would still require the entry in the database to work 🤔

Reply:
With the default salt anybody could tamper with data in the DB because anybody can generate a valid hmac since the shared secret is known.

Reply:
If you decide that there's value in adding a sanity check, would you be willing to make it filterable?

Something like this.

```php
$insecure_salt = apply_filters('wp_cli_login/insecure_salt', 'put your unique phrase here');
```

The reason is because in Bedrock, we use "generateme" and a quick check on GitHub shows that people are using other things like "changeme" "ABCDEFG"  and "" (empty string)

Reply:
> If you decide that there's value in adding a sanity check, would you be willing to make it filterable?
> 
> Something like this.
> 
> ```
> $insecure_salt = apply_filters('wp_cli_login/insecure_salt', 'put your unique phrase here');
> ```
> 
> The reason is because in Bedrock, we use "generateme" and a quick check on GitHub shows that people are using other things like "changeme" "ABCDEFG" and "" (empty string)

Yea, good idea. 

Also, this shows why it's important to extract the "key retrieval" to one place.

You could also add a filter like:

```
$hash_key = apply_filters('wp_cli_login/hash_key', null');

if($hash_key) return $hash_key
```